src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,165 @@
+SG_PRIMITIVE_TYPE(int16_t, TYPE::PT_INT16)
+SG_PRIMITIVE_TYPE(uint16_t, TYPE::PT_UINT16)
+SG_PRIMITIVE_TYPE(int32_t, TYPE::PT_INT32)
+SG_PRIMITIVE_TYPE(uint32_t, TYPE::PT_UINT32)
+SG_PRIMITIVE_TYPE(int64_t, TYPE::PT_INT64)
+SG_PRIMITIVE_TYPE(uint64_t, TYPE::PT_UINT64)
+SG_PRIMITIVE_TYPE(float32_t, TYPE::PT_FLOAT32)
+SG_PRIMITIVE_TYPE(float64_t, TYPE::PT_FLOAT64)
+SG_PRIMITIVE_TYPE(floatmax_t, TYPE::PT_FLOATMAX)
+SG_PRIMITIVE_TYPE(complex128_t, TYPE::PT_COMPLEX128)
+
+#undef SG_PRIMITIVE_TYPE
+
+namespace shogun
+{
+	typedef std::unordered_map<std::type_index, TYPE> typemap;
+
+#define ADD_TYPE_TO_MAP(TYPENAME, TYPE_ENUM)                                   \
+	{std::type_index(typeid(TYPENAME)), TYPE_ENUM},
+	typemap all_types = {
+			ADD_TYPE_TO_MAP(bool, TYPE::PT_BOOL)
+			ADD_TYPE_TO_MAP(char, TYPE::PT_CHAR)
+			ADD_TYPE_TO_MAP(int8_t, TYPE::PT_INT8)
+			ADD_TYPE_TO_MAP(uint8_t, TYPE::PT_UINT8)
+			ADD_TYPE_TO_MAP(int16_t , TYPE::PT_INT16)
+			ADD_TYPE_TO_MAP(uint16_t , TYPE::PT_UINT16)
+			ADD_TYPE_TO_MAP(int32_t , TYPE::PT_INT32)
+			ADD_TYPE_TO_MAP(uint32_t , TYPE::PT_UINT32)
+			ADD_TYPE_TO_MAP(int64_t , TYPE::PT_INT64)
+			ADD_TYPE_TO_MAP(uint64_t , TYPE::PT_UINT64)
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+			ADD_TYPE_TO_MAP(complex128_t, TYPE::PT_COMPLEX128)
+	};
+	typemap non_complex_types = {
+			ADD_TYPE_TO_MAP(bool, TYPE::PT_BOOL)
+			ADD_TYPE_TO_MAP(char, TYPE::PT_CHAR)
+			ADD_TYPE_TO_MAP(int8_t, TYPE::PT_INT8)
+			ADD_TYPE_TO_MAP(uint8_t, TYPE::PT_UINT8)
+			ADD_TYPE_TO_MAP(int16_t , TYPE::PT_INT16)
+			ADD_TYPE_TO_MAP(uint16_t , TYPE::PT_UINT16)
+			ADD_TYPE_TO_MAP(int32_t , TYPE::PT_INT32)
+			ADD_TYPE_TO_MAP(uint32_t , TYPE::PT_UINT32)
+			ADD_TYPE_TO_MAP(int64_t , TYPE::PT_INT64)
+			ADD_TYPE_TO_MAP(uint64_t , TYPE::PT_UINT64)
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+	};
+	typemap real_types = {
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+	};
+	typemap non_integer_types = {
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+			ADD_TYPE_TO_MAP(complex128_t, TYPE::PT_COMPLEX128)
+	};
+
+#undef ADD_TYPE_TO_MAP
+
+	TYPE get_type(const Any& any, typemap map)
+	{
+		std::type_index t = std::type_index(any.type_info());
+		return map[t];
+	}
+	template <typename TypeList, typename Lambda>
+	typename std::enable_if<(not std::is_same<TypeList, Types0>::value), int>::type
+	type_finder(const Any& any, TYPE type, Lambda func)
+	{
+		if (type == sg_primitive_type<typename TypeList::Head>::value)
+		{
+			typename TypeList::Head temporary_type_holder;
+			func(temporary_type_holder);
+			return 1;
```

Comment:
Is there a specific reason for 0 and 1? As opposed to something else?

Reply:
No, could be a boolean, or could be a void function. This is just a way to capture errors, like in my example above we could do this:
```
if (!for_each_type(any, all_types, found_it_lambda)) {
	std::cout << "FAILED" << std::endl;
}
```
Or could just throw an error from within the function if the type isn't found? Or allow a custom lambda to be executed in that situation?

Reply:
mmmh, we could throw an exception (recently added customized exceptions might be worth checking this out). If not, a boolean might be better than the int

Reply:
question is whether there will be situations where no type is found but the caller doensnt know that....
If this does not make sense, exception would be best (with a nice error string) imo

Reply:
I am trying out SG_ERROR, but it won't work here because it seems to only work with classes that inherit from CSObject.

Reply:
Can I just generate an exception inside the function like this:
```
char buffer[256];
sprintf(buffer, "Unsupported type %s", any.type_info().name());
auto msg = std::string(buffer);
throw ShogunException(msg);
``` 

Also I realised that if a type is not in the typemap we can throw an exception immediately. It's only in the case that the type is completely unknown, i.e. not in TYPES, that we will have to go through the list and then throw an exception.

Reply:
@vinx13 could you advise how to best use your new expections?

@gf712 try SG_SERROR for static contexts

Generally, we want to move towards better exceptions, with types. But this is especially true for errors that can be caused by users. This is more an internal error that is caused by a developer...so maybe the SG_SERROR is ok as well

Reply:
Use [`SG_THROW`](https://github.com/shogun-toolbox/shogun/blob/55ebca82b9872cdf5fdb08d5727a3ac92af09a2e/src/shogun/io/SGIO.h#L126) or `SG_STHROW` to throw an exception with custom type. 
The first argument is the type of the exception. The rest of the arguments are the same as `SG_ERROR`
e.g.
`SG_THROW(ShogunException, "...")`

Reply:
Ah cool! Thanks, works as expected!

Reply:
Ah actually no, this gives me a segmentation fault:
```
SG_SERROR("Unsupported type %s", any.type_info().name());
```
which I didn't have when I did this:
```
char buffer[256];
sprintf(buffer, "Unsupported type %s", any.type_info().name());
auto msg = std::string(buffer);
throw ShogunException(msg);
```

Reply:
ah nvm, I just need to add init_shogun() at the start of my example code!